File: Application/VirgilCyptoTests/VC007_PBKDFTests.m
//
//  VC007_PBKDFTests.m
//  VirgilCypto
//
//  Created by Pavel Gorb on 4/27/16.
//  Copyright © 2016 VirgilSecurity. All rights reserved.
//

#import <XCTest/XCTest.h>
#import <Foundation/Foundation.h>

#import "VSSPBKDF.h"

@interface VC007_PBKDFTests : XCTestCase

@end

@implementation VC007_PBKDFTests

- (void)test001_createPBKDF {
    VSSPBKDF *pbkdf = [[VSSPBKDF alloc] initWithSalt:nil iterations:0];
    XCTAssertNotNil(pbkdf, @"VSSPBKDF instance should be created.");
    
    XCTAssertTrue(pbkdf.iterations > 1024, @"VSSPBKDF iterations count should be set to value which is more than 1024.");
    XCTAssertNotNil(pbkdf.salt, @"VSSPBKDF salt should be automatically instantiated.");
    XCTAssertEqual(pbkdf.salt.length, kVSSDefaultRandomBytesSize, @"VSSPBKDF salt size should be equal the default size.");
    
    XCTAssertEqual(pbkdf.algorithm, VSSPBKDFAlgorithmPBKDF2, @"VSSPBKSD algorithm should be properly set to PBKDF2.");
//    pbkdf.algorithm = VSSPBKDFAlgorithmNone;
//    XCTAssertEqual(pbkdf.algorithm, VSSPBKDFAlgorithmNone, @"VSSPBKSD algorithm should be properly changed to None.");
    pbkdf.algorithm = VSSPBKDFAlgorithmPBKDF2;
    XCTAssertEqual(pbkdf.algorithm, VSSPBKDFAlgorithmPBKDF2, @"VSSPBKSD algorithm should be properly set to PBKDF2 again.");
    
    XCTAssertNotEqual(pbkdf.hash, 0, @"VSSPBKSD hash should not be set to 0 by default.");
    XCTAssertNotEqual(pbkdf.hash, VSSPBKDFHashSHA1, @"VSSPBKSD hash should not be set to SHA1 by default.");
    pbkdf.hash = VSSPBKDFHashSHA512;
    XCTAssertEqual(pbkdf.hash, VSSPBKDFHashSHA512, @"VSSPBKSD hash should be properly set to SHA512.");
}

- (void)test002_keyDerivation {
    NSString *password = @"secret";
    size_t keySize = 64;
    NSError *error = nil;
    
    NSData *salt = [VSSPBKDF randomBytesOfSize:0];
    
    VSSPBKDF *pbkdf_a = [[VSSPBKDF alloc] initWithSalt:salt iterations:0];
    NSData *key_a = [pbkdf_a keyFromPassword:password size:keySize error:&error];
    XCTAssertNil(error, @"VSSPBKDF: key should be successfully derived from the password.");
    XCTAssertNotNil(key_a, @"VSSPBKDF: key should contain actual data after the derivation.");
    XCTAssertEqual(key_a.length, keySize, @"VSSPBKDF: key should be generated having the requested size.");
    
    VSSPBKDF *pbkdf_b = [[VSSPBKDF alloc] initWithSalt:salt iterations:0];
    NSData *key_b = [pbkdf_b keyFromPassword:password size:keySize error:&error];
    XCTAssertNil(error, @"VSSPBKDF: key should be successfully derived from the password.");
    XCTAssertNotNil(key_b, @"VSSPBKDF: key should contain actual data after the derivation.");
    XCTAssertEqual(key_b.length, keySize, @"VSSPBKDF: key should be generated having the requested size.");
    
    XCTAssertEqualObjects(key_a, key_b, @"VSSPBKDF: two keys generated independently from the same parameters should match");
    
    VSSPBKDF *pbkdf_c = [[VSSPBKDF alloc] initWithSalt:nil iterations:0];
    NSData *key_c = [pbkdf_c keyFromPassword:password size:keySize error:&error];
    XCTAssertNil(error, @"VSSPBKDF: key should be successfully derived from the password.");
    XCTAssertNotNil(key_a, @"VSSPBKDF: key should contain actual data after the derivation.");
    XCTAssertEqual(key_a.length, keySize, @"VSSPBKDF: key should be generated having the requested size.");
    
    XCTAssertNotEqualObjects(key_a, key_c, @"VSSPBKDF: keys generated with different salt should differ.");
}

- (void)test003_securityChecks {
    VSSPBKDF *pbkdf = [[VSSPBKDF alloc] initWithSalt:nil iterations:0];
    NSError *error = nil;

    BOOL ok = [pbkdf disableRecommendationsCheckWithError:&error];
    XCTAssertTrue(ok, @"VSSPBKDF: security checks should be successfully disabled.");
    XCTAssertNil(error, @"VSSPBKDF: errors should not happen during disabling of the security checks.");
    
    ok = [pbkdf enableRecommendationsCheckWithError:&error];
    XCTAssertTrue(ok, @"VSSPBKDF: security checks should be successfully enabled.");
    XCTAssertNil(error, @"VSSPBKDF: errors should not happen during enabling of the security checks.");
}

- (void)test004_algoritmSettings {
    VSSPBKDF *pbkdf = [[VSSPBKDF alloc] initWithSalt:nil iterations:0];
    
    NSError *error = nil;
    
    pbkdf.algorithm = VSSPBKDFAlgorithmPBKDF2;
    NSData *key_1 = [pbkdf keyFromPassword:@"secret" size:0 error:&error];
    XCTAssertNil(error, @"VSSPBKDF: key should be derived successfully.");
    XCTAssertNotNil(key_1, @"VSSPBKDF: key should be derived from the password.");
}

@end

File: Application/VirgilCryptoSwiftTests/VC007_PBKDFTests.swift
//
//  VC007_PBKDFTests.swift
//  VirgilCypto
//
//  Created by Pavel Gorb on 4/28/16.
//  Copyright © 2016 VirgilSecurity. All rights reserved.
//

import XCTest

class VC007_PBKDFTests: XCTestCase {

    func test001_createPBKDF() {
        let pbkdf = VSCPBKDF(salt: nil, iterations: 0)
     
        XCTAssertNotNil(pbkdf, "VSSPBKDF instance should be created.")
        
        XCTAssertTrue(pbkdf.iterations > 1024, "VSSPBKDF iterations count should be set to value which is more than 1024.")
        XCTAssertNotNil(pbkdf.salt, "VSSPBKDF salt should be automatically instantiated.")
        XCTAssertEqual(pbkdf.salt.count, kVSSDefaultRandomBytesSize, "VSSPBKDF salt size should be equal the default size.")
        
        XCTAssertEqual(pbkdf.algorithm, VSSPBKDFAlgorithm.PBKDF2, "VSSPBKSD algorithm should be properly set to PBKDF2.")

        pbkdf.algorithm = .PBKDF2;
        XCTAssertEqual(pbkdf.algorithm, VSSPBKDFAlgorithm.PBKDF2, "VSSPBKSD algorithm should be properly set to PBKDF2 again.")
        
        XCTAssertNotEqual(pbkdf.hash, VSSPBKDFHash.SHA1, "VSSPBKSD hash should not be set to SHA1 by default.")
        pbkdf.hash = .SHA512;
        XCTAssertEqual(pbkdf.hash, VSSPBKDFHash.SHA512, "VSSPBKSD hash should be properly set to SHA512.")
    }
    
    func test002_keyDerivation() {
        let password = "secret"
        let keySize: size_t = 64
        
        let salt = VSCPBKDF.randomBytes(ofSize: 0)
        
        let pbkdf_a = VSCPBKDF(salt: salt, iterations: 0)
        var key_a: Data? = nil
        do {
            key_a = try pbkdf_a.key(fromPassword: password, size: keySize)
        }
        catch (let error as NSError) {
            XCTFail("VSSPBKDF: key should be derived: \(error.localizedDescription)")
        }
        XCTAssertEqual(key_a!.count, keySize, "VSSPBKDF: key should be generated having the requested size.")
        
        let pbkdf_b = VSCPBKDF(salt: salt, iterations: 0)
        var key_b: Data? = nil
        do {
            key_b = try pbkdf_b.key(fromPassword: password, size: keySize)
        }
        catch (let error as NSError) {
            XCTFail("VSSPBKDF: key should be derived: \(error.localizedDescription)")
        }
        XCTAssertEqual(key_b!.count, keySize, "VSSPBKDF: key should be generated having the requested size.")
        XCTAssertEqual(key_a!, key_b!, "VSSPBKDF: two keys generated independently from the same parameters should match")
        
        let pbkdf_c = VSCPBKDF(salt: nil, iterations: 0)
        var key_c: Data? = nil
        do {
            key_c = try pbkdf_c.key(fromPassword: password, size: keySize)
        }
        catch (let error as NSError) {
            XCTFail("VSSPBKDF: key should be derived: \(error.localizedDescription)")
        }
        XCTAssertEqual(key_c!.count, keySize, "VSSPBKDF: key should be generated having the requested size.")
        XCTAssertNotEqual(key_a!, key_c!, "VSSPBKDF: keys generated with different salt should differ.")
    }
    
    func test003_securityChecks() {
        let pbkdf = VSCPBKDF(salt: nil, iterations: 0)
        
        do {
            try pbkdf.disableRecommendationsCheck()
        }
        catch (let error as NSError) {
            XCTFail("VSSPBKDF: security checks should be disabled: \(error.localizedDescription)")
        }
        
        do {
            try pbkdf.enableRecommendationsCheck()
        }
        catch (let error as NSError) {
            XCTFail("VSSPBKDF: security checks should be enabled: \(error.localizedDescription)")
        }
    }

    func test004_algoritmSettings() {
        let pbkdf = VSCPBKDF(salt: nil, iterations: 0)
        var key: Data? = nil
        
        pbkdf.algorithm = .PBKDF2
        do {
            key = try pbkdf.key(fromPassword: "secret", size: 0)
        }
        catch (let error as NSError) {
            XCTFail("VSSPBKDF: key should be derived successfully: \(error.localizedDescription)")
        }
        XCTAssertNotNil(key, "VSSPBKDF: key should be successfully derived.")
    }
}
